### ISSUE TITLE
Collect data on performance improvements to the dashboard

### ISSUE THREAD

**Issue by shankari** [e-mission/op-admin-dashboard#145]
Plan discussed at meeting

We will use the following environments:
- stage
- ccebikes (large: 86k trips, 108 users)
- smart commute (medium, long-term: 18k trips, 21 users)
- STM community (small: 735 trips, 4 users)

We will access the environments at the following times:
- @JGreenlee 8am MT
- @JGreenlee 12pm MT
- @JGreenlee + @TeachMeTW + @shankari 4pm MT
- @TeachMeTW 8pm MT
- @TeachMeTW 12am MT

Steps to perform:
- Log in aka "cold boot" of the dashboard (initializes to "Overview")
- Data page, UUIDs table, Trip table, Trajectories table
- Maps page with no filters

We will collect data for one week for the baseline and one week for each change.
We will start with reverting the previous batching changes so that we can assess the impact before moving on.

Timeline:
- 21 Oct -25 Oct: baseline (#141 + new PR for reversion)
- 28 Oct - 1 Nov: batching changes (#124, #127, #142)
- 4 Nov - 8 Nov: overview changes (#126 + any bugfixes to #126)

**Comment by shankari** [comment]
to clarify, when I talk about "new PR for reversion", I expect the following

1. `git revert b9b0c347a633a1f44bf697b4cd50e720a279ac09`
2. commit (generate SHA1)
3. add timing (generate SHA2)
4. we determine baseline
5. then, we do `git revert SHA2`; commit
6. `git revert SHA1`; commit
7. then we can test the new code

**Comment by JGreenlee** [comment]
1. `master` has been rolled back to Sep 2, before @TeachMeTW started working on the dashboard
2. what used to be `master` is now called `future`
3. @TeachMeTW is going to add timing to the new `master`. Once that is done, it is considered the "baseline".
4. We will push the baseline to staging

We aim to have the baseline on staging and tested tonight. If good, we'll promote to the 3 prod environments we are testing on, and I will perform the first access time @ 8am MT tomorrow.

If any delay, we will push this back to 8am MT Wednesday.

**Comment by shankari** [comment]
I now see the float error again; we need to cherry-pick that change

```
ERROR:app_sidebar_collapsible:Exception on /_dash-update-component [POST]
Traceback (most recent call last):
  File "/root/miniconda-23.5.2/envs/emission/lib/python3.9/site-packages/flask/app.py", line 2529, in wsgi_app
    response = self.full_dispatch_request()
  File "/root/miniconda-23.5.2/envs/emission/lib/python3.9/site-packages/flask/app.py", line 1825, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/root/miniconda-23.5.2/envs/emission/lib/python3.9/site-packages/flask/app.py", line 1823, in full_dispatch_request
    rv = self.dispatch_request()
  File "/root/miniconda-23.5.2/envs/emission/lib/python3.9/site-packages/flask/app.py", line 1799, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
  File "/root/miniconda-23.5.2/envs/emission/lib/python3.9/site-packages/dash/dash.py", line 1310, in dispatch
    ctx.run(
  File "/root/miniconda-23.5.2/envs/emission/lib/python3.9/site-packages/dash/_callback.py", line 442, in add_context
    output_value = func(*func_args, **func_kwargs)  # %% callback invoked %%
  File "/usr/src/app/app_sidebar_collapsible.py", line 318, in update_store_trips
    df, user_input_cols = query_confirmed_trips(start_date, end_date, timezone)
  File "/usr/src/app/utils/db_utils.py", line 183, in query_confirmed_trips
    df["data.primary_ble_sensed_mode"] = df.ble_sensed_summary.apply(get_max_mode_from_summary)
  File "/root/miniconda-23.5.2/envs/emission/lib/python3.9/site-packages/pandas/core/series.py", line 4771, in apply
    return SeriesApply(self, func, convert_dtype, args, kwargs).apply()
  File "/root/miniconda-23.5.2/envs/emission/lib/python3.9/site-packages/pandas/core/apply.py", line 1123, in apply
    return self.apply_standard()
  File "/root/miniconda-23.5.2/envs/emission/lib/python3.9/site-packages/pandas/core/apply.py", line 1174, in apply_standard
    mapped = lib.map_infer(
  File "pandas/_libs/lib.pyx", line 2924, in pandas._libs.lib.map_infer
  File "/usr/src/app/utils/db_utils.py", line 179, in <lambda>
    get_max_mode_from_summary = lambda md: max(md["distance"], key=md["distance"].get) if len(md["distance"]) > 0 else "INVALID"
TypeError: 'float' object is not subscriptable
```

- Trips trend does not show up
- Trips table is still blank
- Contrary to @TeachMeTW's experience, the UUID table is not blank


<img width="1784" alt="Screenshot 2024-10-21 at 9 22 39 PM" src="https://github.com/user-attachments/assets/f5c97979-8b76-4c68-a27c-7a205b3cbc85">

**Comment by shankari** [comment]
Re-opening since this is not actually done.

**Comment by shankari** [comment]
While collecting data at 4pm MT (3pm PT) today, I ran into some blank tables, likely due to timeouts.

e.g. 
<img width="1786" alt="Screenshot 2024-10-23 at 3 33 44 PM" src="https://github.com/user-attachments/assets/aaec1029-06df-45f4-aa86-0c27b93d2510">

<img width="1786" alt="Screenshot 2024-10-23 at 3 37 06 PM" src="https://github.com/user-attachments/assets/35eb7d48-ceb5-419a-bdd1-3150e928a110">

<img width="1786" alt="Screenshot 2024-10-23 at 3 38 02 PM" src="https://github.com/user-attachments/assets/c683668e-102f-4a0f-963a-ae074f3228eb">

```
Object { message: "Callback error updating tabs-content.children", html: "<html>\r\n<head><title>504 Gateway Time-out</title></head>\r\n<body>\r\n<center><h1>504 Gateway Time-out</h1></center>\r\n<hr><center>nginx</center>\r\n</body>\r\n</html>\r\n" }
[dash_renderer.v2_15_0m1729614367.min.js:2:95440](https://openpath-stage.nrel.gov/admin/_dash-component-suites/dash/dash-renderer/build/dash_renderer.v2_15_0m1729614367.min.js)
    _o https://openpath-stage.nrel.gov/admin/_dash-component-suites/dash/dash-renderer/build/dash_renderer.v2_15_0m1729614367.min.js:2
    ui https://openpath-stage.nrel.gov/admin/_dash-component-suites/dash/dash-renderer/build/dash_renderer.v2_15_0m1729614367.min.js:2
    r https://openpath-stage.nrel.gov/admin/_dash-component-suites/dash/dash-renderer/build/dash_renderer.v2_15_0m1729614367.min.js:2
    r https://openpath-stage.nrel.gov/admin/_dash-component-suites/dash/dash-renderer/build/dash_renderer.v2_15_0m1729614367.min.js:2
    p https://openpath-stage.nrel.gov/admin/_dash-component-suites/dash/dash-renderer/build/dash_renderer.v2_15_0m1729614367.min.js:2
    nt https://openpath-stage.nrel.gov/admin/_dash-component-suites/dash/dash-renderer/build/dash_renderer.v2_15_0m1729614367.min.js:2
    zi https://openpath-stage.nrel.gov/admin/_dash-component-suites/dash/dash-renderer/build/dash_renderer.v2_15_0m1729614367.min.js:2
```

<img width="1790" alt="Screenshot 2024-10-23 at 3 39 02 PM" src="https://github.com/user-attachments/assets/9448672a-a6fc-4c61-b6a8-26f9cd90deea">

Not sure if we are recording that anywhere or can record it given that it is an error in the plotly framework

<img width="1790" alt="Screenshot 2024-10-23 at 3 39 35 PM" src="https://github.com/user-attachments/assets/5cef82fa-8840-473f-99ef-97824f182a4e">


```
Object { message: "Callback error updating card-active-users.children", html: "<html>\r\n<head><title>504 Gateway Time-out</title></head>\r\n<body>\r\n<center><h1>504 Gateway Time-out</h1></center>\r\n<hr><center>nginx</center>\r\n</body>\r\n</html>\r\n" }
[dash_renderer.v2_15_0m1729614367.min.js:2:95440](https://ccebikes-openpath.nrel.gov/admin/_dash-component-suites/dash/dash-renderer/build/dash_renderer.v2_15_0m1729614367.min.js)
    _o https://ccebikes-openpath.nrel.gov/admin/_dash-component-suites/dash/dash-renderer/build/dash_renderer.v2_15_0m1729614367.min.js:2
    ui https://ccebikes-openpath.nrel.gov/admin/_dash-component-suites/dash/dash-renderer/build/dash_renderer.v2_15_0m1729614367.min.js:2
    r https://ccebikes-openpath.nrel.gov/admin/_dash-component-suites/dash/dash-renderer/build/dash_renderer.v2_15_0m1729614367.min.js:2
    r https://ccebikes-openpath.nrel.gov/admin/_dash-component-suites/dash/dash-renderer/build/dash_renderer.v2_15_0m1729614367.min.js:2
    p https://ccebikes-openpath.nrel.gov/admin/_dash-component-suites/dash/dash-renderer/build/dash_renderer.v2_15_0m1729614367.min.js:2
    nt https://ccebikes-openpath.nrel.gov/admin/_dash-component-suites/dash/dash-renderer/build/dash_renderer.v2_15_0m1729614367.min.js:2
    zi https://ccebikes-openpath.nrel.gov/admin/_dash-component-suites/dash/dash-renderer/build/dash_renderer.v2_15_0m1729614367.min.js:2
```

**Comment by TeachMeTW** [comment]
In addition to @shankari's findings; I only had issues with `stage` and `cce-ebikes`; `Smart Commute` and `STM` Loaded with no blanks.

**Comment by JGreenlee** [comment]
As I reported on Teams this morning:

> On ccebikes, UUIDs and Trajectories both timed out after several minutes and failed to display

This held true for the 12pm and 4pm accesses as well. No timeouts on smart commute or stm community.

**Comment by shankari** [comment]
@JGreenlee what about on stage? Are you testing that periodically as well?

**Comment by JGreenlee** [comment]
I tested Stage just now. UUIDs eventually loaded. Trips, Trajectories, and Heatmap were empty – I don't think it's due to a timeout, I think it is due to a lack of data in the last week

Starting tomorrow I'll access Stage along with the others.

**Comment by TeachMeTW** [comment]
8 PM MT:

- CCEbike did not load active users box
- Smart Commute did  not load uuids
- Stage did not load uuids or trajectories or trips (might need to add date change to this test?)

**Comment by TeachMeTW** [comment]
12 AM MT:

- CCEbike did not load active users box
- Smart Commute now loads uuids
- CCEbike long load times in trajectories
- Stage loads uuids
- Stage no trips or trajectories

**Comment by JGreenlee** [comment]
> On ccebikes, UUIDs and Trajectories both timed out after several minutes and failed to display
> 
> This held true for the 12pm and 4pm accesses as well. No timeouts on smart commute or stm community.

> I tested Stage just now. UUIDs eventually loaded. Trips, Trajectories, and Heatmap were empty – I don't think it's due to a timeout, I think it is due to a lack of data in the last week
> 
> Starting tomorrow I'll access Stage along with the others.

8am MT was consistent with my findings from yesterday.

The active users box is indeed missing on ccebikes; it was likely missing yesterday as well but I didn't notice.

**Comment by JGreenlee** [comment]
I'm going to make sure my staging phones are on and connected to Wi-Fi today so that we have at least a couple trips on `stage` to populate Trips, Trajectories and the map

**Comment by TeachMeTW** [comment]
Tested: 4 PM MT 

- Home page functioned normally for all
- CCEbike UUIDs did NOT load
- Stage still no trips or trajectories; seems not added yet
- CCEbike slow on trajectories

**Comment by TeachMeTW** [comment]
@shankari I discussed with @JGreenlee in regards to the analysis, he suggested this workflow; what are your thoughts?

```
I recommend performing exploratory analysis+viz before removing outliers so that you know the data you're working with, you can see the outliers and identify them.
 
You will have dozens of fine-grained measures for execution time of small tasks. I think you should try to extract a couple higher-level measures that are specifically related to the changes you made.
I suggest something like i) total time to load all figures on the overview and ii) time to load the UUIDs table
 
Then generate descriptives+ charts for each high-level measure ("basic stuff" like mean execution time by week, dataset size)
You could generate these for fine-grained measures too, for exploratory purposes, but probably won't make it in the paper.
 
If I remember stats, I think the main analysis you want to perform is probably a two-way repeated measures ANOVA. We are mostly interested in 2 IVs:
The week (week 1 = baseline; week 2 = first batch of improvements; week 3 = both batches of improvements).
Dataset size (small = stm community, medium = smart commute, large = ccebikes)
With the DV being execution time for the high level tasks
 
This will allow you to test for:
Main effect of week (Did performance change significantly as we rolled out improvements?)
Main effect of dataset size (Does performance significantly degrade with size of the dataset?)
Interaction effect (Does the performance change differently depending on the size of the dataset? (highlight this for scalability)

I don't see what a correlation matrix would be useful for.
I don't see a need for modeling or ML unless (i) you think the time of day has a significant effect that you want to model or (ii) you want a regression model to predict how the system would scale to even larger programs (larger than ccebikes). But I suggest trying to keep your analysis fairly small-scale.
```

**Comment by shankari** [comment]
I was a little late starting today because I forgot (I should put a reminder in my calendar). I started at closer to 4:30pm MT.

**Comment by shankari** [comment]
@TeachMeTW @JGreenlee is your mentor, so I would take his advice 😄 
Seriously, if there is something that JGreenlee is unclear about and needs my help, I am happy to step in.
But otherwise, I trust Jack's judgement, and am happy to review results once they are ready and provide feedback that way.

Completely agree on "no ML", EDA != ML

**Comment by TeachMeTW** [comment]
8 PM MT:

- Home Page loaded normally albeit slow on ccebikes
- CCEbike did not load uuids
- Still no trajectories or trips on stage

**Comment by shankari** [comment]
> Still no trajectories or trips on stage

Interesting. I saw one trip and its associated trajectory on stage (presumably from Jack's phone)

**Comment by TeachMeTW** [comment]
12 AM MT:

- Home page loaded normally
- CCEbike did not load uuids
- Still no trajectories or trips on stage

**Comment by JGreenlee** [comment]
I saw my trip on staging yesterday and this morning.
Behavior is still fairly consistent on my end, except this morning I noticed that Trajectories on ccebikes loaded successfully instead of timing out.

**Comment by JGreenlee** [comment]
> Seriously, if there is something that JGreenlee is unclear about and needs my help, I am happy to step in. But otherwise, I trust Jack's judgement, and am happy to review results once they are ready and provide feedback that way.

@shankari For the record, I had suggested that @TeachMeTW get your advice in addition to mine since most of my experience with study design + statistical analysis has been in a different domain

**Comment by TeachMeTW** [comment]
4 PM MT:

- Only STM loaded uuids
- Trajectories for CCEbike taking 15+ min/permanently loading

**Comment by shankari** [comment]
I forgot today too and am just loading now. @TeachMeTW did you do this at 4pm PT or 4pm MT?

**Comment by TeachMeTW** [comment]
@shankari I started at 3:15 PST, did not finish until 4 PST; mainly ccebike not finishing loading at all

**Comment by TeachMeTW** [comment]
```python
%env DB_HOST=mongodb://localhost/openpath_prod_ca_ebike
import emission.core.get_database as edb
import emission.storage.timeseries.aggregate_timeseries as esta
import emission.storage.timeseries.builtin_timeseries as estb
import emission.core.get_database as gdb
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

collection = gdb.get_timeseries_db()
```

    env: DB_HOST=mongodb://localhost/openpath_prod_ca_ebike



```python
# Define the query filter
query_filter = {"metadata.key": "stats/dashboard_time"}

# Define the projection to include specific fields and exclude '_id'
# keep everything else
projection = {
    "_id": 0,
    "user_id": 1,
    "metadata.key": 1,
    "metadata.platform": 1,
    "metadata.time_zone": 1,
    "metadata.write_ts": 1,
    "metadata.write_local_dt": 1,
    "metadata.write_fmt_time": 1,
    "data.name": 1,
    "data.ts": 1,
    "data.reading": 1
}

# Execute the find query
cursor = collection.find(filter=query_filter, projection=projection)

# Convert the cursor to a list of documents
data = list(cursor)

# Display the number of documents fetched
print(f"Number of documents fetched: {len(data)}")
```

    Number of documents fetched: 221



```python
# Normalize the list of JSON documents into a flat table
df = pd.json_normalize(data)

# Display the first few rows of the DataFrame
print("\nDataFrame Head:")
print(df.head())
```

    
    DataFrame Head:
      user_id          metadata.key metadata.platform  metadata.write_ts  \
    0    None  stats/dashboard_time            server       1.729555e+09   
    1    None  stats/dashboard_time            server       1.729555e+09   
    2    None  stats/dashboard_time            server       1.729555e+09   
    3    None  stats/dashboard_time            server       1.729555e+09   
    4    None  stats/dashboard_time            server       1.729555e+09   
    
        metadata.time_zone  metadata.write_local_dt.year  \
    0  America/Los_Angeles                          2024   
    1  America/Los_Angeles                          2024   
    2  America/Los_Angeles                          2024   
    3  America/Los_Angeles                          2024   
    4  America/Los_Angeles                          2024   
    
       metadata.write_local_dt.month  metadata.write_local_dt.day  \
    0                             10                           21   
    1                             10                           21   
    2                             10                           21   
    3                             10                           21   
    4                             10                           21   
    
       metadata.write_local_dt.hour  metadata.write_local_dt.minute  \
    0                            16                              49   
    1                            16                              49   
    2                            16                              49   
    3                            16                              49   
    4                            16                              49   
    
       metadata.write_local_dt.second  metadata.write_local_dt.weekday  \
    0                              50                                0   
    1                              50                                0   
    2                              50                                0   
    3                              50                                0   
    4                              50                                0   
    
      metadata.write_local_dt.timezone           metadata.write_fmt_time  \
    0              America/Los_Angeles  2024-10-21T16:49:50.932304-07:00   
    1              America/Los_Angeles  2024-10-21T16:49:50.943540-07:00   
    2              America/Los_Angeles  2024-10-21T16:49:50.951867-07:00   
    3              America/Los_Angeles  2024-10-21T16:49:50.953847-07:00   
    4              America/Los_Angeles  2024-10-21T16:49:50.954911-07:00   
    
                                               data.name       data.ts  \
    0  admin/db_utils/query_uuids/logging_and_query_i...  1.729555e+09   
    1  admin/db_utils/query_uuids/convert_query_resul...  1.729555e+09   
    2    admin/db_utils/query_uuids/dataframe_processing  1.729555e+09   
    3              admin/db_utils/query_uuids/total_time  1.729555e+09   
    4  admin/db_utils/df_to_filtered_records/filter_d...  1.729555e+09   
    
       data.reading  
    0      0.459791  
    1      5.287625  
    2      5.427958  
    3     22.177542  
    4      0.000584  



```python
# Drop rows with missing 'data.reading' or 'data.ts'
df.dropna(subset=['data.reading', 'data.ts'], inplace=True)

# Display the cleaned DataFrame
print("\nCleaned DataFrame:")
print(df.head())
```

    
    Cleaned DataFrame:
      user_id          metadata.key metadata.platform  metadata.write_ts  \
    0    None  stats/dashboard_time            server       1.729555e+09   
    1    None  stats/dashboard_time            server       1.729555e+09   
    2    None  stats/dashboard_time            server       1.729555e+09   
    3    None  stats/dashboard_time            server       1.729555e+09   
    4    None  stats/dashboard_time            server       1.729555e+09   
    
        metadata.time_zone  metadata.write_local_dt.year  \
    0  America/Los_Angeles                          2024   
    1  America/Los_Angeles                          2024   
    2  America/Los_Angeles                          2024   
    3  America/Los_Angeles                          2024   
    4  America/Los_Angeles                          2024   
    
       metadata.write_local_dt.month  metadata.write_local_dt.day  \
    0                             10                           21   
    1                             10                           21   
    2                             10                           21   
    3                             10                           21   
    4                             10                           21   
    
       metadata.write_local_dt.hour  metadata.write_local_dt.minute  \
    0                            16                              49   
    1                            16                              49   
    2                            16                              49   
    3                            16                              49   
    4                            16                              49   
    
       metadata.write_local_dt.second  metadata.write_local_dt.weekday  \
    0                              50                                0   
    1                              50                                0   
    2                              50                                0   
    3                              50                                0   
    4                              50                                0   
    
      metadata.write_local_dt.timezone           metadata.write_fmt_time  \
    0              America/Los_Angeles  2024-10-21T16:49:50.932304-07:00   
    1              America/Los_Angeles  2024-10-21T16:49:50.943540-07:00   
    2              America/Los_Angeles  2024-10-21T16:49:50.951867-07:00   
    3              America/Los_Angeles  2024-10-21T16:49:50.953847-07:00   
    4              America/Los_Angeles  2024-10-21T16:49:50.954911-07:00   
    
                                               data.name       data.ts  \
    0  admin/db_utils/query_uuids/logging_and_query_i...  1.729555e+09   
    1  admin/db_utils/query_uuids/convert_query_resul...  1.729555e+09   
    2    admin/db_utils/query_uuids/dataframe_processing  1.729555e+09   
    3              admin/db_utils/query_uuids/total_time  1.729555e+09   
    4  admin/db_utils/df_to_filtered_records/filter_d...  1.729555e+09   
    
       data.reading  
    0      0.459791  
    1      5.287625  
    2      5.427958  
    3     22.177542  
    4      0.000584  



```python
# Convert 'metadata.write_fmt_time' to datetime objects
df['write_fmt_time'] = pd.to_datetime(df['metadata.write_fmt_time'], errors='coerce')

# Check for any parsing errors
missing_times = df['write_fmt_time'].isna().sum()
if missing_times > 0:
    print(f"\nWarning: {missing_times} records have invalid 'write_fmt_time' and will be dropped.")
    df = df.dropna(subset=['write_fmt_time'])

# Extract hour from 'write_fmt_time'
df['hour'] = df['write_fmt_time'].dt.hour

# Display the DataFrame with the new 'hour' column
print("\nDataFrame with Hour:")
print(df[['write_fmt_time', 'hour']].head())
```

    
    DataFrame with Hour:
                        write_fmt_time  hour
    0 2024-10-21 16:49:50.932304-07:00    16
    1 2024-10-21 16:49:50.943540-07:00    16
    2 2024-10-21 16:49:50.951867-07:00    16
    3 2024-10-21 16:49:50.953847-07:00    16
    4 2024-10-21 16:49:50.954911-07:00    16



```python
# Group by 'hour' and 'data name' and compute average reading and count
grouped = df.groupby(['hour', 'data.name']).agg(
    avg_reading=pd.NamedAgg(column='data.reading', aggfunc='mean'),
    count=pd.NamedAgg(column='data.reading', aggfunc='count')
).reset_index()

# Display the grouped DataFrame
print("\nGrouped Data by Hour and Function:")
print(grouped.head())
```

    
    Grouped Data by Hour and Function:
       hour                                          data.name  avg_reading  count
    0    16  admin/data/populate_datatable/check_dataframe_...     0.000750      1
    1    16     admin/data/populate_datatable/create_datatable     0.621875      1
    2    16           admin/data/populate_datatable/total_time     1.835542      1
    3    16         admin/data/render_content/handle_uuids_tab  1831.903042      1
    4    16  admin/data/render_content/prepare_final_datafr...     3.536000      1



```python
# limit count
# top_n = 10
# top_functions = grouped.groupby('function')['count'].sum().nlargest(top_n).index
top_functions = grouped.groupby('data.name')['count'].sum().index
heatmap_data = grouped[grouped['data.name'].isin(top_functions)].pivot("data.name", "hour", "avg_reading")

# Set the size of the heatmap
plt.figure(figsize=(20, 12))

# Create the heatmap with annotations
sns.heatmap(heatmap_data, annot=True, fmt=".2f", cmap="YlGnBu", linewidths=.5)

# Add title and labels
plt.title('Average Readings by Function and Hour of Day')
plt.xlabel('Hour of Day')
plt.ylabel('Function')

# Show the plot
plt.tight_layout()
plt.show()
```

    /var/folders/8j/284cnbzn2tzbfljhtsr17gsx2490jk/T/ipykernel_54541/3454771667.py:5: FutureWarning: In a future version of pandas all arguments of DataFrame.pivot will be keyword-only.
      heatmap_data = grouped[grouped['data.name'].isin(top_functions)].pivot("data.name", "hour", "avg_reading")



![image](https://github.com/user-attachments/assets/ee18f8dc-f8ed-49eb-bd87-2de2a72d7048)

    



```python
specific_function = "admin/db_utils/add_user_stats/total_time"
# Filter the grouped data for the specific function
func_data = grouped[grouped['data.name'] == specific_function]

# Check if the function exists in the data
if not func_data.empty:
    # Set the plot size
    plt.figure(figsize=(12, 6))
    
    # Create a bar chart
    sns.barplot(x='hour', y='avg_reading', data=func_data, palette='viridis')
    
    # Add title and labels
    plt.title(f'Average Readings by Hour for Function: {specific_function}')
    plt.xlabel('Hour of Day')
    plt.ylabel('Average Reading')
    
    # Show the plot
    plt.tight_layout()
    plt.show()
else:
    print(f"\nFunction '{specific_function}' not found in the data.")
```

    /var/folders/8j/284cnbzn2tzbfljhtsr17gsx2490jk/T/ipykernel_54541/4202144101.py:11: FutureWarning: 
    
    Passing `palette` without assigning `hue` is deprecated and will be removed in v0.14.0. Assign the `x` variable to `hue` and set `legend=False` for the same effect.
    
      sns.barplot(x='hour', y='avg_reading', data=func_data, palette='viridis')



    
![image](https://github.com/user-attachments/assets/7395d6b6-cc35-4bfe-b0a2-c6bbe1377465)

    



```python
#top_n = 10
#top_functions = grouped.groupby('function')['count'].sum().nlargest(top_n).index
top_functions = grouped.groupby('data.name')['count'].sum().index
line_plot_data = grouped[grouped['data.name'].isin(top_functions)].pivot(index='hour', columns='data.name', values='avg_reading')

# Set the plot size
plt.figure(figsize=(20, 12))

# Plot each function's average reading over hours
for function in line_plot_data.columns:
    plt.plot(line_plot_data.index, line_plot_data[function], marker='o', label=function)

# Add title and labels
plt.title('Average Readings by Hour for Each Function')
plt.xlabel('Hour of Day')
plt.ylabel('Average Reading')

# Add legend
plt.legend(title='Function', bbox_to_anchor=(1.05, 1), loc='upper left')

# Show the plot
plt.tight_layout()
plt.show()
```


![image](https://github.com/user-attachments/assets/c257acbb-724f-49f5-ae82-261016a62c3a)
    



```python
# Find the hour with the highest average reading for each function
peak_hours = grouped.loc[grouped.groupby('data.name')['avg_reading'].idxmax()][['data.name', 'hour', 'avg_reading']]

# Display the peak hours
print("\nPeak Hours for Each Function:")
print(peak_hours)
```

    
    Peak Hours for Each Function:
                                                data.name  hour  avg_reading
    0   admin/data/populate_datatable/check_dataframe_...    16     0.000750
    1      admin/data/populate_datatable/create_datatable    16     0.621875
    2            admin/data/populate_datatable/total_time    16     1.835542
    3          admin/data/render_content/handle_uuids_tab    16  1831.903042
    4   admin/data/render_content/prepare_final_datafr...    16     3.536000
    ..                                                ...   ...          ...
    66   admin/home/update_card_trips/generate_trips_card    16    17.149708
    67            admin/home/update_card_trips/total_time    16    46.285646
    68        update_card_users/calculate_number_of_users    16     0.005062
    69               update_card_users/generate_user_card    16    22.212916
    70                       update_card_users/total_time    16    36.257896
    
    [71 rows x 3 columns]

**Comment by TeachMeTW** [comment]
- CCEBike did NOT load active users
- Only stage and stm loaded uuids

**Comment by shankari** [comment]
@TeachMeTW did you collect data today? We started this week of data collection on Wednesday, so should continue collecting data with this version today and tomorrow.

**Comment by TeachMeTW** [comment]
@shankari No I did not; I will get to that now. I got confused/thought it was only until friday

**Comment by TeachMeTW** [comment]
9 PM MT:
- CCEbike no active users
- CCEbike no uuids

**Comment by TeachMeTW** [comment]
12 AM MT:
- All loaded home page normally
- CCEbike no uuids

**Comment by TeachMeTW** [comment]
3 MT:

- Only STM loaded UUIDs
- The rest are not loading in a reasonable time

**Comment by TeachMeTW** [comment]
12 AM MT:
- CCEbike no uuid

**Comment by TeachMeTW** [comment]
6 PM MT:
Tested Update on Staging
- Loading updates work
- Home Page normal
- UUIDs loaded chunks normally
- Tested table page retention: Works
- Trajectories switcher worked albeit slow
- Maps Work

**Comment by JGreenlee** [comment]
I started the next phase of data collection this morning, repeating the same steps as before. I observed batch loading of UUIDs working. I waited for all UUIDs to finish loading before I proceeded to the "Trips" tab.

No 'active minutes' on ccebikes

**Comment by JGreenlee** [comment]
On ccebikes, I found that the initial load of the UUIDs tab still took a long time to initialize – even though it now shows only 10 users at first. Then it loads 10 users every batch at a constant rate, so it took a while for all 111 users to be shown.

**Comment by TeachMeTW** [comment]
@shankari  @JGreenlee 

Given the current setup where only 10 users are loaded initially and subsequent batches load at a steady rate (10 users per 24000 ms), would it be possible to improve testing speed by only loading a subset of batches? This would reduce overall wait time during testing, although we may lose some data points and miss potential issues that might appear in later batches. Do you think this trade-off would be acceptable, or is it important to retain the full load for thorough testing?

**Comment by TeachMeTW** [comment]
4 PM MT:
- STM Community loaded normally
- Smart Commute and CCEbike loaded batches way later than expected 24000ms but eventually loaded.
- Trajectories and switching keys take a long time

**Comment by TeachMeTW** [comment]
9 PM MT:
- Everything works, slightly faster than last time

**Comment by TeachMeTW** [comment]
4 AM MT:
- Same as last entry

**Comment by TeachMeTW** [comment]
@shankari @JGreenlee CCEbike Graphs -- for round 1 of changes (NOT BASELINE)
![image](https://github.com/user-attachments/assets/919cbad2-5837-482d-a074-1579d98bf741)
![image](https://github.com/user-attachments/assets/824613df-e564-4515-bd57-59fe306f9e20)
![image](https://github.com/user-attachments/assets/fa0250ee-abe2-4468-b77c-bec1eed96d07)

**Comment by TeachMeTW** [comment]
More graphs:
![image](https://github.com/user-attachments/assets/d04d4e36-2060-47ed-af0f-0b814718b4a1)
![image](https://github.com/user-attachments/assets/2baa13cc-988b-4f2e-b1b4-4f5ad35e5ace)

**Comment by shankari** [comment]
Results do seem good. But I would like some explanation and narrative around this.
Are these results only for staging, or also for other environments? What is the explanation for them, etc?
I would suggest filling out the results section of your report with the charts and the associated narrative.
The goal is not to make visually appealing charts; the goal is to make it easier for people to understand your results

**Comment by shankari** [comment]
Methodology:
- block diagram of admin dashboard
- instrumented the various sections
- logged in at some cadence; collected data for a week
- made improvements
- logged in at another cadence; collected data for another week
- compared before and after

Implementation and results:
- found that the majority of the time was spent in these section
- improvement 1
   - we realized that the issue was that we were loading all the data at the same time
   - started batching
   - issues with implementing batching
   - improvement results
- improvement 2
   - now max time was spend in ....
   - can precompute and cache
   - issues with implementing precomputation and caching
   - improvement results (TBD
- summary and overall results

**Comment by TeachMeTW** [comment]
@shankari until stage is reuploaded here are my insights:

# Function, Execution Time (seconds)

## Lowest 80%
```
data.name,data.reading
admin/data/update_dropdowns_trips/determine_hidden_columns_and_label,1.947862946675444e-06
admin/data/populate_datatable/check_dataframe_type,2.046613198896547e-06
admin/data/update_store_trajectories/prepare_store_data,2.060361875919625e-06
update_card_users/calculate_number_of_users,3.4012816479948702e-06
admin/home/update_card_trips/calculate_number_of_trips,3.9885911499605635e-06
admin/data/update_sub_tab/retrieve_and_process_data,3.998180697332525e-06
admin/home/generate_plot_trips_trend/convert_iso_to_date_only,4.4479934529036935e-06
admin/db_utils/clean_location_data/total_time,3.3919877825276354e-05
admin/db_utils/query_uuids/logging_and_query_initiation,8.000350707152393e-05
admin/data/render_content/handle_demographics_tab,0.00013015080367040355
admin/db_utils/query_demographics/organize_survey_keys,0.00015084072795919687
admin/home/generate_card/generate_card_layout,0.00016447972178438623
admin/home/find_last_get/convert_to_uuid_objects,0.00023495605481522425
admin/db_utils/query_trajectories/convert_date_range_to_timestamps,0.00023877198247646448
admin/db_utils/query_confirmed_trips/convert_date_range_to_timestamps,0.0003273542747809275
admin/home/generate_barplot/update_layout_with_title,0.00037205597217397103
admin/home/compute_trips_trend/extract_date,0.0005048959373316789
admin/home/get_number_of_active_users/calculate_active_users,0.0006949753012685548
admin/db_utils/query_confirmed_trips/rename_columns,0.0006985895775609866
admin/home/update_card_active_users/convert_to_dataframe,0.0007020320802453965
admin/home/generate_plot_sign_up_trend/convert_to_dataframe,0.0007046631796926748
admin/data/update_sub_tab/filter_columns,0.0007352904097691256
admin/home/compute_sign_up_trend/convert_to_datetime,0.0011078711812857723
admin/home/compute_trips_trend/group_by_and_calculate_counts,0.0019007130519995649
admin/data/update_sub_tab/convert_to_dataframe,0.001956743379448906
admin/db_utils/query_confirmed_trips/process_binary_and_named_columns,0.0026572991253141524
admin/db_utils/query_uuids/dataframe_processing,0.0028467976734427793
admin/db_utils/add_user_stats/retrieve_user_profile_data,0.003226581826238717
admin/home/compute_sign_up_trend/group_by_and_calculate_counts,0.0032378398798039703
admin/home/generate_plot_trips_trend/convert_to_dataframe,0.0032821491783935294
admin/db_utils/df_to_filtered_records/filter_dataframe_if_needed,0.003408925289243042
admin/home/compute_trips_trend/convert_to_datetime,0.003680147491926935
admin/db_utils/query_confirmed_trips/process_coordinates_and_modes,0.005306827750419394
admin/db_utils/query_confirmed_trips/expand_and_filter_user_inputs,0.009418276858057305
admin/db_utils/query_uuids/convert_query_result_to_dataframe,0.010353045917858253
admin/home/generate_card/total_time,0.011176629554714
admin/data/update_dropdowns_trips/total_time,0.012006994434029496
admin/db_utils/query_demographics/create_dataframes,0.015817682580615968
admin/home/update_card_active_users/generate_active_users_card,0.01920611160023818
admin/db_utils/query_demographics/process_dataframes,0.019244120024051795
admin/data/render_content/handle_uuids_tab,0.019926290027797222
admin/db_utils/query_confirmed_trips/humanize_distance_and_duration,0.022384994969278132
admin/home/update_card_trips/generate_trips_card,0.022472899088667043
update_card_users/generate_user_card,0.025483410393239044
admin/home/compute_sign_up_trend/total_time,0.02933595221123246
admin/data/update_sub_tab/populate_datatable,0.0318839150521247
admin/data/render_content/total_time,0.03323825279854329
admin/home/compute_trips_trend/total_time,0.035378578287908684
admin/data/populate_datatable/create_datatable,0.03774038002808034
admin/home/generate_barplot/generate_barplot_with_data,0.039363626396305536
admin/home/generate_plot_sign_up_trend/compute_sign_up_trend,0.0400892436870786
admin/home/generate_plot_trips_trend/compute_trips_trend,0.044780456077680116
admin/home/update_card_trips/total_time,0.046317080631406105
admin/db_utils/df_to_filtered_records/convert_to_dict_of_records,0.04962066936440856
admin/db_utils/query_uuids/total_time,0.04976970487803555
admin/data/render_content/handle_trips_tab,0.0507602873720316
admin/home/generate_barplot/initialize_empty_barplot,0.051887884061879025
update_card_users/total_time,0.0561006364000381
admin/data/populate_datatable/total_time,0.05679313991567768
admin/db_utils/add_user_stats/get_last_trip_timestamp,0.06127944804549133
admin/db_utils/add_user_stats/get_first_trip_timestamp,0.06184159415772293
admin/db_utils/add_user_stats/count_labeled_trips,0.06342866654713583
admin/db_utils/df_to_filtered_records/total_time,0.07345906148523829
admin/data/update_sub_tab/total_time,0.0753064348988725
admin/db_utils/query_demographics/query_data,0.08909172358757392
admin/home/generate_plot_trips_trend/generate_barplot,0.12236394326804578
admin/home/generate_barplot/total_time,0.1288078651732151
admin/home/generate_plot_sign_up_trend/generate_barplot,0.1545202970718507
admin/db_utils/query_trajectories/add_mode_string,0.1633187119957438
admin/db_utils/query_demographics/total_time,0.17353401138483432
admin/data/render_content/prepare_final_dataframe_and_return,0.2047745715041297
admin/home/generate_plot_trips_trend/total_time,0.21455150090016262
admin/home/generate_plot_sign_up_trend/total_time,0.2335722325064364
admin/data/update_store_trajectories/filter_records,0.4412800445715402
admin/db_utils/query_trajectories/process_dataframe_columns,0.5254711637183209
admin/db_utils/add_user_stats/count_total_trips,0.8359979697176627
admin/db_utils/query_confirmed_trips/retrieve_aggregate_time_series,2.383731212208139
admin/db_utils/query_confirmed_trips/total_time,2.399663755555122
admin/db_utils/add_user_stats/process_user,2.4734303609223955
admin/db_utils/add_user_stats/get_last_server_call_timestamp,2.890578399274938
```
## Top 20%
```
data.name,data.reading
admin/home/update_card_active_users/total_time,380.5705834423147
admin/home/update_card_active_users/calculate_active_users,380.51765034232466
admin/home/get_number_of_active_users/total_time,380.50897508538844
admin/home/get_number_of_active_users/find_last_get_entries,380.49126240247415
admin/home/find_last_get/total_time,380.48344891932277
admin/home/find_last_get/query_timeseries_db,380.4540818235763
admin/data/render_content/handle_trajectories_tab,364.7788547482385
admin/data/update_store_trajectories/total_time,364.5321865298467
admin/data/update_store_trajectories/query_trajectories,364.06441769682766
admin/db_utils/query_trajectories/total_time,364.0559167172793
admin/data/render_content/total_time,352.24918736368954
admin/data/render_content/handle_uuids_tab,350.33841586602665
admin/db_utils/add_user_stats/total_time,350.3015870457756
admin/db_utils/query_trajectories/retrieve_entries,213.97728912887717
admin/db_utils/query_trajectories/convert_entries_to_dataframe,149.3421039039905
admin/db_utils/add_user_stats/process_batches,44.96146263926428
admin/db_utils/add_user_stats/processing_loop_stage,44.95224149511222
admin/db_utils/add_user_stats/get_last_trip_timestamp,44.94357279768172
admin/db_utils/add_user_stats/get_last_server_call_timestamp,23.636255698262932
admin/db_utils/add_user_stats/process_user,20.8464001387978
admin/db_utils/query_confirmed_trips/retrieve_aggregate_time_series,8.48705214093778
admin/db_utils/query_confirmed_trips/total_time,8.38841827082712
admin/db_utils/add_user_stats/count_total_trips,6.84013815707495
```

**Comment by TeachMeTW** [comment]
# Function-wise:
## Bottom 80%:
```
function,data.reading
add_user_stats,350.3015870457756
clean_location_data,3.3919877825276354e-05
compute_sign_up_trend,0.02933595221123246
compute_trips_trend,0.035378578287908684
df_to_filtered_records,0.07345906148523829
generate_barplot,0.1288078651732151
generate_card,0.011176629554714
generate_plot_sign_up_trend,0.2335722325064364
generate_plot_trips_trend,0.21455150090016262
populate_datatable,0.05679313991567768
query_confirmed_trips,4.53850465386655
query_demographics,0.17353401138483432
query_uuids,0.04976970487803555
render_content,332.586646515266
update_card_trips,0.046317080631406105
update_dropdowns_trips,0.012006994434029496
update_sub_tab,0.0753064348988725
```


## Top 20%:
```
function,data.reading
find_last_get,380.48344891932277
get_number_of_active_users,380.50897508538844
query_trajectories,364.0559167172793
update_card_active_users,380.5705834423147
update_store_trajectories,364.5321865298467
```

**Comment by shankari** [comment]
@TeachMeTW I suggested removing the bottom 80% of stats **for the pipeline changes** not the admin dashboard. The admin dashboard is only launched when a user logs in, so the additional data storage is minimal. The pipeline runs every hour, so the additional data storage is substantial.

I am not opposed to (eventually) dropping the bottom 80% of timing stats in the admin dashboard, but that is not blocking the push to production of the pipeline timing changes.

**Comment by TeachMeTW** [comment]
@shankari please reupload the staging data snapshot; the one from tuesday became corrupted

**Comment by shankari** [comment]
I also see `add_user_stats,350.3015870457756` in the bottom 80% which looks wrong. What does the number there represent?

**Comment by TeachMeTW** [comment]
> I also see `add_user_stats,350.3015870457756` in the bottom 80% which looks wrong. What does the number there represent?

That number is in **seconds** of execution time

**Comment by shankari** [comment]
After merging https://github.com/e-mission/e-mission-server/pull/993 and https://github.com/e-mission/e-mission-server/pull/1005 and https://github.com/e-mission/op-admin-dashboard/pull/153, I logged in to staging.

And the initial "Overview" load was still pretty slow. But we were still loading the UUIDs in a batch when I went to the data table.

So presumably the "Overview" load didn't load all the UUIDs? Because if it did, we should just have cached and reused that data, right? So then what is the overview loading?

The functionality seems to be working correctly, and we can collect data next week, but I think that this can benefit from some cleanup.

**Comment by shankari** [comment]
For the record, while cleaning up https://github.com/e-mission/e-mission-server/pull/1005, you can also split the user stats into pipeline-dependent and pipeline-independent. The pipeline-dependent stats can stay as the last stage in the pipeline, and the pipeline-independent stats can be outside the early return. This will allow us to not waste time recomputing the pipeline-dependent stats when the pipeline wasn't run.

The current structure of `run_intake_pipeline_for_user` may make it hard to test the pipeline-independent code; feel free to refactor/restructure if that makes it cleaner.

**Comment by shankari** [comment]
Deployed to the three production environments as well.
Happy testing and make sure to write down your testing time/results here.

**Comment by JGreenlee** [comment]
> And the initial "Overview" load was still pretty slow. But we were still loading the UUIDs in a batch when I went to the data table.
> 
> So presumably the "Overview" load didn't load all the UUIDs? Because if it did, we should just have cached and reused that data, right? So then what is the overview loading?

I believe the Overview page does create the entire list of UUIDs.

On the Data page UUIDs table, we are loading **stats** for those UUIDs. We already have the full list of UUIDs by the time we get to the Data table, and we split that list into slices of 10 UUIDs and load the stats for one slice at a time.

**Comment by JGreenlee** [comment]
Accessed all 4 dashboards ~9am MT. Nothing appears significantly different than before.
The Overview is slow. UUIDs table batching works but each batch takes ~25 seconds.

The Trips and Trajectories tabs did not take as long as I remembered. Maybe it's simply because it's winter and there is less travel in the last week.

I didn't notice anything broken on any of the dashboards.

There are some recent trips on `stage` but no recent trips on `stm-community`.

**Comment by JGreenlee** [comment]
> UUIDs table batching works but each batch takes ~25 seconds.

This is likely happening faster computationally due to offloading the user stats computation to the pipeline, but we are not observing the effect because of the 20 second interval between batches.
I think we will need to look at stored `stats/dashboard_time` entries to see how much faster it is.

**Comment by JGreenlee** [comment]
Loading the overview seemed slower today across the board.
I also notice that the "Active Users" box failed to load on ccebikes and smart-commute. I don't believe it has consistently been that way

**Comment by TeachMeTW** [comment]
Same results for me as well

**Comment by TeachMeTW** [comment]
Too add to this, today seems better for smart commut but ccebikes still fails on active users

**Comment by shankari** [comment]
Why are you changing to the 2022-2025 range? Is this something you did earlier?

**Comment by TeachMeTW** [comment]
@shankari it is an arbitrary range to test the date picker and to see if the data would updated. Previously, I only changed the month so perhaps I should stay consistent with that. I chose 2022 - 2025 as it could encompass most data.

**Comment by JGreenlee** [comment]
We should stick to what we were doing before (which I thought was keeping the filters untouched)

If our previous measurements were taken with 1 week of data and our current measurements are taken with 2-3 years of data, we can't make comparisons

**Comment by TeachMeTW** [comment]
@JGreenlee Understood I will rerun the suite without changing the filters

**Comment by shankari** [comment]
@JGreenlee  @TeachMeTW https://github.com/e-mission/op-admin-dashboard/pull/156 is now on production across the board. You can start testing against all the canonical production environments

**Comment by TeachMeTW** [comment]
@shankari 

<img width="1724" alt="Image" src="https://github.com/user-attachments/assets/9506b6d1-70ac-417d-a457-d7f4d4e15318" />

tested on CCEbike, as you said trajectories took a while to load but eventually did/

**Comment by TeachMeTW** [comment]
@shankari 
Test Stage 2/12/25:

Test Process:

- Open Overview Page
- Open Data page
- Visit UUIDs Tab
- Visit Trips Tab
- Visit Demographics Tab
- Visit Trajectories Tab
- Changed to background/location
- Open Maps Page
- Change Map type to Density
- Change modes to Unlabeled

Test Notes:
- Instant Overview Page loaded
- UUIDs loaded normally and as expected on batches
- Trips loaded instantly
- Demographics the same
**- Trajectories quite slow; does this need batch loading? This is for both recreated and background location -- especially slow on background location. Like a couple minutes slow.**
- Map Performed Normally

Also tested the new stats and added to op dash:

https://github.com/e-mission/op-admin-dashboard/pull/164

**Comment by shankari** [comment]
@TeachMeTW are we testing this in production? I don't see any updates here, and the changes have been on production for a couple of weeks.

**Comment by TeachMeTW** [comment]
> [@TeachMeTW](https://github.com/TeachMeTW) are we testing this in production? I don't see any updates here, and the changes have been on production for a couple of weeks.

The 4 production environments? Yes I still am, no new updates really same as https://github.com/e-mission/op-admin-dashboard/issues/145#issuecomment-2655667168 regardless of the environment, the trajectories tab is quite slow. But nothing new to report other than that.

**Comment by shankari** [comment]
I think at least a weekly update would be good

**Comment by TeachMeTW** [comment]
Tested this week:

Test Process:

Open Overview Page
Open Data page
Visit UUIDs Tab
Visit Trips Tab
Visit Demographics Tab
Visit Trajectories Tab
Changed to background/location
Open Maps Page
Change Map type to Density
Change modes to Unlabeled


Test Notes:

Instant Overview Page loaded
UUIDs loaded normally and as expected on batches
Trips loaded instantly
Demographics the same
Trajectories normal for stage
Map Performed Normally
